### PR TITLE
ADR: Standardize storage target layout

### PR DESCRIPTION
**Pull request by rabernat**
cc @pangeo-forge/dev-team

### CODE REVIEW

**Comment by martindurant** [comment]
This effectively describes a path-as-metadata scheme, so that a listing of some storage acts like a catalogue; except that the metadata is stored within the output files (maybe). Is that the intent? Is listing these entities on some catalogue service a separate and orthogonal part of the framework?
Should there be a particular place to store **how** to load a given path as a dataset and (maybe) how datasets relate to one-another?

**Comment by rabernat** [comment]
> This effectively describes a path-as-metadata scheme, so that a listing of some storage acts like a catalogue; except that the metadata is stored within the output files (maybe). Is that the intent?

Yes, the intent is that the listing of the bucket tells you **what datasets from what recipes exist in the store**. For more metadata, you would have to either
- peek into the datasets
- cross-reference the recipe `meta.yaml`

I imagine a catalog service will eventually do both.

> Is listing these entities on some catalogue service a separate and orthogonal part of the framework?

Separate but related. The hope is that the Bakery can store things according to this path convention, and the catalog service can act independently, crawling and indexing the buckets as necessary.

> Should there be a particular place to store **how** to load a given path as a dataset and (maybe) how datasets relate to one-another?

"How to load a dataset" is key for intake integration. Ideally it would map cleanly to the `dataset_type` key (e.g. `zarr`). Can you think of any more information needed to store? If so, where do we store it?

How datasets relate to one another is a very interesting question. I imagine we could put that in `meta.yaml` somehow.

**Comment by martindurant** [comment]
> "How to load a dataset" is key for intake integration. Ideally it would map cleanly to the dataset_type key (e.g. zarr). Can you think of any more information needed to store? If so, where do we store it?

In the current world, most datasets seem to need extra arguments when passing to xarray, and I don't suppose that will go away. An fsspec-reference-file, for example, will not store the arguments (`anon=True`, `requester_pays=True`, ...) for the target storage internally.

The BAD but simple solution would be to store an intake YAML alongside each dataset. I would think that kind of thing most naturally belongs in a database. "Database" of course might include a cataloguing standard like STAC somehow.
How such defs map to languages or implementations is another matter.

**Comment by rabernat** [comment]
Thanks for the comments Martin!

At our last meeting we decided that each ADR would require at least one review / approval from @pangeo-forge/dev-team. Could someone please review and / or merge this?

File: doc/adr/0003-standardize-storage-target-layout.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+/pangeo-us-west-2/pangeo-forge/noaa_oisst/avhrr_only.zarr
+```
+
+`prefix` can be any valid characters for the object store. We assume that
+the bakery manager may not have control over this, e.g. if one organization is
+sharing a single bucket among many teams.
+
+For the other three keys, we will use a restricted set of characters.
+- the `/` character is interpreted as a path separator and can't be used
+  **Question: do we want to allow `recipe_name` to include "sub-directories"**?
```
Comment (rabernat) [inline doc/adr/0003-standardize-storage-target-layout.md:68]: Any thoughts on this question? Will we allow recipe feedstocks to define their own nested sub-hierarchies of datasets?

File: doc/adr/0003-standardize-storage-target-layout.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+- non `/` characters must be valid python identifiers:
+  - digits `0-9`
+  - lowercase letters `a-z` (lowercase keeps path names clean and pythonic)
+  - underscores `_`
```
Comment (rabernat) [inline doc/adr/0003-standardize-storage-target-layout.md:72]: A lot of real datasets have `.` characters in them, e.g. for version numbers. Do we want to consider allowing `.`? Is it important that the names be valid python identifiers, or is that too restrictive?

**Pull request merged**